### PR TITLE
direct io write support

### PR DESCRIPTION
**Pull request by lightmark**
rocksdb direct io support

```
[gzh@dev11575.prn2 ~/rocksdb] ./db_bench -benchmarks=fillseq --num=1000000
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
RocksDB:    version 5.0
Date:       Wed Nov 23 13:17:43 2016
CPU:        40 * Intel(R) Xeon(R) CPU E5-2660 v2 @ 2.20GHz
CPUCache:   25600 KB
Keys:       16 bytes each
Values:     100 bytes each (50 bytes after compression)
Entries:    1000000
Prefix:    0 bytes
Keys per prefix:    0
RawSize:    110.6 MB (estimated)
FileSize:   62.9 MB (estimated)
Write rate: 0 bytes/second
Compression: Snappy
Memtablerep: skip_list
Perf Level: 1
WARNING: Assertions are enabled; benchmarks unnecessarily slow
------------------------------------------------
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
DB path: [/tmp/rocksdbtest-112628/dbbench]
fillseq      :       4.393 micros/op 227639 ops/sec;   25.2 MB/s

[gzh@dev11575.prn2 ~/rocksdb] ./db_bench -use_direct_writes -benchmarks=fillseq --num=1000000
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
RocksDB:    version 5.0
Date:       Wed Nov 23 13:18:10 2016
CPU:        40 * Intel(R) Xeon(R) CPU E5-2660 v2 @ 2.20GHz
CPUCache:   25600 KB
Keys:       16 bytes each
Values:     100 bytes each (50 bytes after compression)
Entries:    1000000
Prefix:    0 bytes
Keys per prefix:    0
RawSize:    110.6 MB (estimated)
FileSize:   62.9 MB (estimated)
Write rate: 0 bytes/second
Compression: Snappy
Memtablerep: skip_list
Perf Level: 1
WARNING: Assertions are enabled; benchmarks unnecessarily slow
------------------------------------------------
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
DB path: [/tmp/rocksdbtest-112628/dbbench]
fillseq      :       7.621 micros/op 131222 ops/sec;   14.5 MB/s

### CODE REVIEW

**Comment by siddontang** [comment]
Seem that the performance is reduced with direct IO write.

**Comment by lightmark** [comment]
Yes, because I omitted some optimization options for direct io writes.
```
[gzh@dev11575.prn2 ~/rocksdb] ./db_bench -benchmarks=fillseq -use_direct_writes=1 -skip_table_builder_flush=1 -new_table_reader_for_compaction_inputs=1 -compaction_readahead_size=10485760 --writable_file_max_buffer_size=10485760 --num=1000000
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
RocksDB:    version 5.0
Date:       Mon Nov 28 11:08:13 2016
CPU:        40 * Intel(R) Xeon(R) CPU E5-2660 v2 @ 2.20GHz
CPUCache:   25600 KB
Keys:       16 bytes each
Values:     100 bytes each (50 bytes after compression)
Entries:    1000000
Prefix:    0 bytes
Keys per prefix:    0
RawSize:    110.6 MB (estimated)
FileSize:   62.9 MB (estimated)
Write rate: 0 bytes/second
Compression: Snappy
Memtablerep: skip_list
Perf Level: 1
WARNING: Assertions are enabled; benchmarks unnecessarily slow
------------------------------------------------
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
DB path: [/tmp/rocksdbtest-112628/dbbench]
fillseq      :       3.942 micros/op 253666 ops/sec;   28.1 MB/s
```

**Comment by IslamAbdelRahman** [comment]
@lightmark, can you please your diff to trigger the tests ? I don't see any tests in Phabricator

**Comment by IslamAbdelRahman** [comment]
@lightmark, looks like windows is broken .. can you please fix and rebase

**Comment by IslamAbdelRahman** [comment]
as discussed offline, let's try to avoid `enforce_buffered_io`

File: port/win/env_win.cc [hunk 1]
```diff
@@ -737,10 +737,10 @@ std::string WinEnvIO::TimeToString(uint64_t secondsSince1970) {
 EnvOptions WinEnvIO::OptimizeForLogWrite(const EnvOptions& env_options,
   const DBOptions& db_options) const {
   EnvOptions optimized = env_options;
-  optimized.use_mmap_writes = false;
   optimized.bytes_per_sync = db_options.wal_bytes_per_sync;
-  optimized.use_os_buffer =
-    true;  // This is because we flush only whole pages on unbuffered io and
+  optimized.use_mmap_writes = false;
+  optimized.use_direct_writes = false;
+  // This is because we flush only whole pages on unbuffered io and
```
Comment (IslamAbdelRahman) [inline port/win/env_win.cc]: looks to me that this comment need to move up

File: port/win/io_win.h [hunk 2]
```diff
@@ -437,30 +420,27 @@ class WinWritableFile : private WinFileData,
...
-    const EnvOptions& options);
+                  const EnvOptions& options);
 
   ~WinRandomRWFile() {}
 
-  // Indicates if the class makes use of unbuffered I/O
+  // Indicates if the class makes use of direct I/O
   // If false you must pass aligned buffer to Write()
-  virtual bool UseOSBuffer() const override;
+  virtual bool UseDirectIO() const override;
```
Comment (IslamAbdelRahman) [inline port/win/io_win.h]: `If true you must ...`

File: tools/db_bench_tool.cc [hunk 3]
```diff
@@ -2937,9 +2938,10 @@ class Benchmark {
     options.optimize_filters_for_hits = FLAGS_optimize_filters_for_hits;
 
     // fill storage options
-    options.allow_os_buffer = FLAGS_bufferedio;
     options.allow_mmap_reads = FLAGS_mmap_read;
     options.allow_mmap_writes = FLAGS_mmap_write;
+    options.use_direct_reads = FLAGS_use_direct_reads;
+    options.use_direct_writes = FLAGS_use_direct_writes;
```
Comment (IslamAbdelRahman) [inline tools/db_bench_tool.cc]: these 2 lines look to me are a duplicate of 2746

File: util/db_options.cc [hunk 4]
```diff
@@ -152,12 +152,14 @@ void ImmutableDBOptions::Dump(Logger* log) const {
   Header(log,
          "            Options.manifest_preallocation_size: %" ROCKSDB_PRIszt,
          manifest_preallocation_size);
-  Header(log, "                        Options.allow_os_buffer: %d",
-         allow_os_buffer);
   Header(log, "                       Options.allow_mmap_reads: %d",
          allow_mmap_reads);
   Header(log, "                      Options.allow_mmap_writes: %d",
          allow_mmap_writes);
+  Header(log, "                       Options.use_direct_reads: %d",
```
Comment (IslamAbdelRahman) [inline util/db_options.cc]: looks to me this part is a duplicate of 126 -> 133

File: util/io_posix.cc [hunk 5]
```diff
@@ -187,7 +187,7 @@ Status PosixSequentialFile::Read(size_t n, Slice* result, char* scratch) {
       s = IOError(filename_, errno);
     }
   }
-  if (!use_os_buffer_) {
+  if (use_direct_io_) {
```
Comment (IslamAbdelRahman) [inline util/io_posix.cc]: Why are we still calling Fadvise

File: util/io_posix.cc [hunk 6]
```diff
@@ -325,7 +325,7 @@ Status PosixRandomAccessFile::Read(uint64_t offset, size_t n, Slice* result,
     // An error: return a non-ok status
     s = IOError(filename_, errno);
   }
-  if (!use_os_buffer_) {
+  if (use_direct_io_) {
```
Comment (IslamAbdelRahman) [inline util/io_posix.cc]: Same comment

File: util/io_posix.cc [hunk 7]
```diff
@@ -680,6 +684,8 @@ PosixWritableFile::~PosixWritableFile() {
 }
 
 Status PosixWritableFile::Append(const Slice& data) {
+  assert(!direct_io_ ||
```
Comment (IslamAbdelRahman) [inline util/io_posix.cc]: we should also have an assert in PosixWritableFile::PositionedAppend

File: util/options.cc [hunk 8]
```diff
@@ -336,11 +336,12 @@ void DBOptions::Dump(Logger* log) const {
...
     Header(log, "  Options.recycle_log_file_num: %" ROCKSDB_PRIszt,
            recycle_log_file_num);
-    Header(log, "       Options.allow_os_buffer: %d", allow_os_buffer);
-    Header(log, "      Options.allow_mmap_reads: %d", allow_mmap_reads);
     Header(log, "      Options.allow_fallocate: %d", allow_fallocate);
+    Header(log, "      Options.allow_mmap_reads: %d", allow_mmap_reads);
     Header(log, "     Options.allow_mmap_writes: %d", allow_mmap_writes);
     Header(log, "      Options.use_direct_reads: %d", use_direct_reads);
+    Header(log, "     Options.use_direct_writes: %d", use_direct_writes);
+    Header(log, "      Options.use_direct_reads: %d", use_direct_reads);
```
Comment (IslamAbdelRahman) [inline util/options.cc]: looks to me that these lines are duplicates

File: util/options.cc [hunk 9]
```diff
@@ -366,12 +367,14 @@ void DBOptions::Dump(Logger* log) const {
...
          manifest_preallocation_size);
-    Header(log, "                         Options.allow_os_buffer: %d",
-        allow_os_buffer);
     Header(log, "                        Options.allow_mmap_reads: %d",
         allow_mmap_reads);
     Header(log, "                       Options.allow_mmap_writes: %d",
         allow_mmap_writes);
+    Header(log, "                        Options.use_direct_reads: %d",
+           use_direct_reads);
+    Header(log, "                       Options.use_direct_writes: %d",
```
Comment (IslamAbdelRahman) [inline util/options.cc]: also duplicate

File: util/options_helper.h [hunk 10]
```diff
@@ -186,12 +186,12 @@ static std::unordered_map<std::string, OptionTypeInfo> db_options_type_info = {
     {"use_direct_reads",
      {offsetof(struct DBOptions, use_direct_reads), OptionType::kBoolean,
       OptionVerificationType::kNormal, false, 0}},
+    {"use_direct_writes",
+     {offsetof(struct DBOptions, use_direct_writes), OptionType::kBoolean,
+      OptionVerificationType::kNormal, false, 0}},
     {"allow_2pc",
      {offsetof(struct DBOptions, allow_2pc), OptionType::kBoolean,
       OptionVerificationType::kNormal, false, 0}},
-    {"allow_os_buffer",
```
Comment (IslamAbdelRahman) [inline util/options_helper.h]: Should be `kDeprecated` instead of deleted

File: util/env_posix.cc [hunk 11]
```diff
@@ -729,14 +758,14 @@ class PosixEnv : public Env {
       const EnvOptions& env_options) const override {
     EnvOptions optimized = env_options;
     optimized.use_mmap_writes = false;
+    optimized.use_direct_writes = false;
```
Comment (lightmark) [inline util/env_posix.cc]: I edited env_posix options here @IslamAbdelRahman

**Comment by IslamAbdelRahman** [comment]
@lightmark, can you please re-import to trigger the tests after your most recent changes

File: db/db_impl.cc [hunk 12]
```diff
@@ -255,11 +255,11 @@ static Status ValidateOptions(
         "More than four DB paths are not supported yet. ");
   }
 
-  if (db_options.allow_mmap_reads && !db_options.allow_os_buffer) {
+  if (db_options.allow_mmap_reads && db_options.use_direct_reads) {
```
Comment (IslamAbdelRahman) [inline db/db_impl.cc]: I think you missed this comment

Comment (IslamAbdelRahman) [inline db/db_impl.cc]: you did not, my bad :D

**Comment by IslamAbdelRahman** [comment]
Reviewed 47 of 48 files at r1.
Review status: 47 of 48 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[db/db_test.cc, line 3407 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZUDa6ISinX-5YSaW2G:-KZUDa6ISinX-5YSaW2H:bp1yd3x) ([raw file](https://github.com/facebook/rocksdb/blob/874aeb932b63030b76adf96743d8967d33e1aaec/db/db_test.cc#L3407)):*
> ```cpp
>   ASSERT_OK(TryReopen(options));
> 
>   // make check -j64 usually run on tmpfs
> ```

Can you please explain more ?

---

*[port/win/io_win.h, line 78 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZUJJJmBitildt266XG:-KZUJJJmBitildt266XH:byryhr7) ([raw file](https://github.com/facebook/rocksdb/blob/874aeb932b63030b76adf96743d8967d33e1aaec/port/win/io_win.h#L78)):*
> ```objc
>   // If perf evaluation finds this to be a problem, we can look into
>   // implementing this.
>   const bool use_direct_io_;
> ```

the comment here is not correct anymore

---

*[util/env_posix.cc, line 325 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZUMDRNpT12f7GEPOi3:-KZUMDRNpT12f7GEPOi4:b-mb7emo) ([raw file](https://github.com/facebook/rocksdb/blob/874aeb932b63030b76adf96743d8967d33e1aaec/util/env_posix.cc#L325)):*
> ```cpp
>     int fd = -1;
> 
>     int flags = O_TRUNC;
> ```

This looks wrong to me, we dont truncate when reusing a writable file

---

*[util/env_posix.cc, line 389 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZULbpyRoX0WmDlvHru:-KZULbpyRoX0WmDlvHrv:bxx73ua) ([raw file](https://github.com/facebook/rocksdb/blob/874aeb932b63030b76adf96743d8967d33e1aaec/util/env_posix.cc#L389)):*
> ```cpp
>   }
> 
>   virtual Status NewRandomRWFile(const std::string& fname,
> ```

What about `NewRandomRWFile`

---

*[util/log_write_bench.cc, line 37 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZUL6Sk9_tfqm0rbaVo:-KZUL6Sk9_tfqm0rbaVp:b-wfpkog) ([raw file](https://github.com/facebook/rocksdb/blob/874aeb932b63030b76adf96743d8967d33e1aaec/util/log_write_bench.cc#L37)):*
> ```cpp
>   std::string file_name = test::TmpDir() + "/log_write_benchmark.log";
>   Env* env = Env::Default();
>   EnvOptions env_options = env->OptimizeForLogWrite(EnvOptions());
> ```

now we are ignoring FLAGS_bytes_per_sync;

---

*[utilities/backupable/backupable_db.cc, line 1157 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZUKILrlr1lW9QU-lR5:-KZUKILsdVl1S60kGdc4:bq74hux) ([raw file](https://github.com/facebook/rocksdb/blob/dfb6fe6755bfce3b73d572f15ab9fa081bc15270/utilities/backupable/backupable_db.cc#L1157)):*
> ```cpp
>   EnvOptions env_options;
>   env_options.use_mmap_writes = false;
>   env_options.use_os_buffer = false;
> ```

let's enable direct IO for backupable db

---

*[utilities/backupable/backupable_db_test.cc, line 386 at r1](https://reviewable.io:443/reviews/facebook/rocksdb/1564#-KZUGsMISYBbACY1UZow:-KZUGsMISYBbACY1UZox:b-lbh5p3) ([raw file](https://github.com/facebook/rocksdb/blob/874aeb932b63030b76adf96743d8967d33e1aaec/utilities/backupable/backupable_db_test.cc#L386)):*
> ```cpp
>     EnvOptions env_options;
>     env_options.use_mmap_writes = false;
>     env_options.use_direct_writes = false;
> ```

just wondering, why did we add this line ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/facebook/rocksdb/1564)*

**Comment by IslamAbdelRahman** [comment]
Looks great, mostly minor comments

**Comment by IslamAbdelRahman** [comment]
can you check the compilation error as well ?

**Pull request closed without merging**